vocabulary-draft/continuous-delivery.md

```diff
@@ -0,0 +1,114 @@
+
+Due the dynamic nature of Pipelines, most of actual work needs to be queued to happen in a distributed way, hence Queued events are added. Adopters can choose to ignore these events if they don't apply to their use cases. 
+
+- **PipelineRun Queued**: a PipelineRun has been schedule to run
+- **PipelineRun Started**: a PipelineRun has started and it is running
+- **PipelineRun Finished**: a PipelineRun has finished it execution, the event will contain the finished status, success, error or failure
+
+Each pipeline is defined as a set of Tasks to be performed in sequence, hence tracking this sequence might be important for some cases. A TaskRun is an instance of the Task defined inside the pipeline, as you can expect multiple execution of the pipelines (each a PipelineRun) you can also expect multiple execution of the Tasks, for that reason we use TaskRun to refer to one of these instances.  
+
+- **TaskRun Started**: a TaskRun inside a PipelineRun has started.
+- **TaskRun Finished**: a TaskRun inside a PipelineRun has finished. 
+
+Pipeline Events MUST include the following attributes:
+- **Event Type**: the type is restricted to include `CD.Pipeline**` prefix. For example `CD.Pipeline.PipelineRun.Queued` or `CD.Pipeline.Tests.Started`
+- **PipelineRun Id**: unique identifier for a pipeline execution
+- **Pipeline Name**: unique identifier for the pipeline, not for the instance. A pipeline can have multiple instances/runs.  
+- **PipelineRun Status**: current status of the PipelineRun at the time when the event was emitted. If the pipeline is finished, this attribute should reflect if it finished successfully or if there was an error on the execution. Possible statuses: [Running, Finished, Error]
+
+Optional attributes: 
+- **PipelineRun URL**: URL to locate where  pipeline instances are running
+- **PipelineRun Errors**: error field to indicate possible compilation, test, build and package errors.
+
+## Product Composition Events
+
+When the products delivered to users can be, or needs to be, broken down into components (hardware, software, configuration files etc.), the creation of versions of said components and the compositoning of components into products or product subsystems may warrant the use of separate events detailed in this section.
+
+Product Composition Events MUST include the following attributes:
+
+- **Event Type**: the type is restricted to include `CD.ProductComposition` prefix. For example `CD.ProductComposition.ComponentVersionCreated` or `CD.ProductComposition.CompositionCreated`
+
+### Component Version Created
```

Comment:
@erkist I am really interested in this and how a Component is different from an Artifact.. are we talking about the same things? 

The term product confuses me here.. because you might be building an Open Source library and not actually a product, I still believe that we are talking about something that is really close to an Artifact, in the sense of that you can have a Software Artifact (code, binaries, docker images) a Documentation Artifact (word document) or a hardware artifact (raspberry pi).

I wonder if there is another connotation to call this a Component vs Artifact that pushes us to use Component instead.

Reply:
I agree that it is very close to an artifact, and there was a note saying roughly the same thing in the old HackMD document.

There are two things I consider components but not artifacts, as currently described:
1. Several artifacts (i.e. several "loose" artifacts, not several artifacts packed together in an archive)
2. Hardware

So, a component version could be "physically represented" by only one artifact, but it could also be several artifacts or a hardware component version (and to clarify: not an instance of a hardware (e.g. "my laptop") but a version of hardware (e.g. "Thinkpad X290 rev 44").

So, to be able to answer the question "Have we tested this firmware upgrade on a Thinkpad X290 rev 44?", we need to be able to build compositions that include both firmware and hardware. If we can find another way of doing so that reduces the overlap between Component and Artifact, that could be a way forward.

I think Eiffel proposes a solution where a composition can include artifacts directly, without going via component versions, which could be one way of solving it.